File: Python/Spiral Matrix.py
class Solution:
    # @param matrix, a list of lists of integers
    # @return a list of integers
    def spiralOrder(self, matrix):
        m = len(matrix)
        if m == 0 :
            return []
        if m == 1:
            return matrix[0]

        n = len(matrix[0])
        if n == 1:
            return [matrix[i][0] for i in range(m)]

        ret = []

        up = 0
        down = m-1
        left = 0
        right = n-1

        while up <=  down and left <=  right :
            if left == right + 1:break
            for i in range( left, right + 1):
                ret.append(matrix[up][i])
            up += 1

            if up == down + 1 :break 
            for i in range(up, down + 1):
                ret.append(matrix[i][right])
            right -= 1

            if right == left - 1: break
            for i in range(right, left-1, -1):
                ret.append(matrix[down][i])
            down -= 1

            if down == up - 1: break 
            for i in range(down , up - 1, -1):
                ret.append(matrix[i][left])
            left += 1
        return ret

File: Java/src/Spiral Matrix.java
import java.util.ArrayList;

public class Solution
{
  public List<Integer> spiralOrder(int[][] matrix)
  {
    List<Integer> ret = new ArrayList<>();
    int m = matrix.length;
    if (m == 0)
      return ret;
    int n = matrix[0].length;

    int up = 0;
    int down = m - 1;
    int left = 0;
    int right = n - 1;

    while (up <= down && left <= right)
    {
      for (int j = left; j <= right; j++)
        ret.add(matrix[up][j]);
      up++;

      for (int j = up; j <= down; j++)
        ret.add(matrix[j][right]);
      right--;

      if (up <= down)
      {
        for (int j = right; j >= left; j--)
          ret.add(matrix[down][j]);
      }
      down--;

      if (left <= right)
      {
        for (int j = down; j >= up; j--)
          ret.add(matrix[j][left]);
      }
      left++;

    }
    return ret;
  }
}
